kernel/work.c

```diff
@@ -0,0 +1,952 @@
+/*
+ * Copyright (c) 2020 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+/**
+ * @file
+ *
+ * Second generation work queue implementation
+ */
+
+#include <kernel.h>
+#include <kernel_structs.h>
+#include <wait_q.h>
+#include <spinlock.h>
+#include <errno.h>
+#include <sys/scheduler.h>
+#include <sys/printk.h>
+
+/* Lock to protect the internal state of all work items, and
+ * pending_cancels.
+ *
+ * No queue locks may be held when this is taken, but a queue lock may
+ * be taken while this is held.
+ */
+static struct k_spinlock lock;
```

Comment:
global locks like this are somewhat legacy, is it possible to move the granularity of this lock to  the individual work items, or the relevant workqueue they are on?

Reply:
I don't think you really answered my question.

> Unfortunately, no. This protects the flags of individual work items across complex states checks/changes that can't be performed with a single atomic action, at times when the work item might not be associated with a queue. 

I didn't say we leave all concurrency to the queue object. If the intention is to protect flags in work items, why not introduce a lock into the work item itself instead of having a global spinlock here? So the queue has a lock, and the work items have themselves locks, instead of a global lock for all the work items.

> Similarly it protects queue flags which have to be manipulated while the work flags are protected, and vice-versa.

You can use multiple spinlocks in a critical section as long as they are always acquired and released in the same order. Can you provide more detail on why you think this is not possible?